### DISCUSSION TITLE
get_attribute swallows tracebacks for KeyError/AttributeError

### DISCUSSION THREAD

**Discussion by johtso** [encode/django-rest-framework#7745] [Potential Issue]
[`get_attribute` currently catches `KeyError`s and `AttributeError`s when trying to retrieve fields](https://github.com/tomchristie/django-rest-framework/blob/0ca1145456b998ab1aee7083e006d9fb111b1d0a/rest_framework/fields.py#L300), assuming that those errors resulted from the top level attribute getting or key getting.

The problem with this is that if _anything else_ inside a property causes one of those errors, the traceback is swallowed and it's a nightmare to debug!

Is there maybe a way to make sure that the exception doesn't come from deeper down?

Contrived illustration:

> E   AttributeError: Got AttributeError when attempting to get a value for field `my_prop` on serializer `WidgetSerializer`.
> E   The serializer field might be named incorrectly and not match any attribute or key on the `Widget` instance.
> E   Original exception text was: 'something_entirely_unrelated' object has no attribute 'some_attribute'.

**Comment by johtso** [comment]
Hmm, I think I have a possible solution.

First of all this error message logic could be moved to [the base `get_attribute` function](https://github.com/tomchristie/django-rest-framework/blob/0ca1145456b998ab1aee7083e006d9fb111b1d0a/rest_framework/fields.py#L67). It doesn't seem right that the error handling logic is split between the `get_attribute` function and the method. One catches `ObjectDoesNotExist`, the other catches `KeyError` and `AttributeError`. Why not do that all in one place.

Secondly, instead of catching `KeyError` and `AttributeError` exceptions, we could use `getitem` and `getattr` with _defaults_ and respond to the default being returned.

Then we can be precise about the issue we're trying to catch, and avoid confusing a exception thrown by code triggered by the _getattr_ or _getitem_ / property call, for an error retrieving the key/attr itself.

How does that sound?

**Comment by sirpengi** [comment]
I'm having this issue as well, just here to add some more background.

This is somewhat related to #2598, but instead of callables raising an attribute exception this is on property fields.  Here's a reproduction:

``` python
"""
requirements.txt:
Django==1.8.2
djangorestframework==3.1.3
"""

from django.conf import settings
SETTINGS = dict(
    DATABASES = {},
    DEBUG=True,
    TEMPLATE_DEBUG=True,
    ROOT_URLCONF = None,
)
settings.configure(**SETTINGS)


from rest_framework import serializers

class ExampleSerializer(serializers.Serializer):
        example_property = serializers.CharField(read_only=True)
        example_fail_property = serializers.CharField(read_only=True)

class ExampleInstance(object):
        @property
        def example_property(self):
                return "foobar"

        @property
        def example_fail_property(self):
                raise AttributeError('property failed')

serializer = ExampleSerializer(ExampleInstance())
print(serializer.data.items()) # ItemsView({'example_property': 'foobar'})
```

**Comment by xordoquy** [comment]
@johtso my first thought is that is feels wrong to have a `KeyError` or `AttributeError`.
I'm going to keep this opened to dig a bit more into the issue before I make a final decision.

**Comment by sirpengi** [comment]
Just another note as I'm digging further into this: this problem only manifests for `read_only=True` fields.  The exceptions otherwise bubble up properly.

**Comment by johtso** [comment]
Just got bitten by this again.. any more thoughts on this problem/solution?

**Comment by drew-moore** [comment]
just got bit. ouch.

**Comment by lovelydinosaur** [comment]
Not obvious how to proceed with this. Thoughts/suggestions welcome.

**Comment by chidg** [comment]
Just for any googlers arriving here trying to solve a KeyError on get_attribute, it may be due to accessing serializer.data in a situation where validation has already been done. In that case you should be using serializer.validated_data.

http://stackoverflow.com/questions/32236958/keyerror-on-relation-field-in-django-rest

**Comment by rpkilby** [comment]
Something to think about later - #5600 could be adapted to fix this.

**Comment by cgi1** [comment]
@chidg: Good point; still appears in 2018.

**Comment by munsu** [comment]
Was caught by this. Some other package I rely on for a property had a non-obvious breaking change. My serializer tests failed instantly -- just dropped the field from `serializer.data` with no warning/error. Spent half a day tracing the source (thought it was something else breaking the dict comparison in my tests).

A warning log at the least would have been immensely helpful.

**Comment by AndersonSMed** [comment]
Just found out about this discussion and the very same thing still happens today. I also created a simple code snippet that illustrates this behavior:
```python3
from rest_framework import serializers

class ExampleSerializer(serializers.Serializer):
    title = serializers.CharField(read_only=True)

class ExampleObject:
    @property
    def title(self):
        return self.foo

obj = ExampleObject()
serializer = ExampleSerializer()

# Doesn't trigger any error
serializer.to_representation(obj)

# Triggers an AttributeError
obj.title
```
